Dev2

jianhong opened the pull request:
<!--
# nf-core/hicar pull request

Many thanks for contributing to nf-core/hicar!

Please fill in the appropriate checklist below (delete whatever is not relevant).
These are the most common things requested on pull requests (PRs).

Remember that PRs should be made against the dev branch, unless you're preparing a pipeline release.

Learn more about contributing: [CONTRIBUTING.md](https://github.com/nf-core/hicar/tree/master/.github/CONTRIBUTING.md)
-->

## PR checklist

- [x] This comment contains a description of changes (with reason).
- [x] If you've fixed a bug or added code that should be tested, add tests!
  - [x] If you've added a new tool - have you followed the pipeline conventions in the [contribution docs](https://github.com/nf-core/hicar/tree/master/.github/CONTRIBUTING.md)
  - [ ] If necessary, also make a PR on the nf-core/hicar _branch_ on the [nf-core/test-datasets](https://github.com/nf-core/test-datasets) repository.
- [x] Make sure your code lints (`nf-core lint`).
- [x] Ensure the test suite passes (`nextflow run . -profile test,docker --outdir <OUTDIR>`).
- [x] Usage Documentation in `docs/usage.md` is updated.
- [x] Output Documentation in `docs/output.md` is updated.
- [x] `CHANGELOG.md` is updated.
- [x] `README.md` is updated (including new tool citations and authors/contributors).

nf-core-bot committed:
Template update for nf-core/tools version 2.4

jianhong committed:
add support for MseI.

jianhong committed:
Merge remote-tracking branch 'upstream/dev' into dev

jianhong committed:
Merge remote-tracking branch 'upstream/nf-core-template-merge-2.4' into dev

jianhong committed:
fix the issue of multiqc_config file

jianhong committed:
fix a issue in README.

jianhong committed:
add TAD, AB compartments, APA analysis

jianhong committed:
change the pipeline to use different tools

jianhong committed:
update workflow image.

jianhong committed:
re-orgnize the juicer tools modules.

jianhong committed:
add multiple scripts for homer

jianhong committed:
update scripts for hicdcplus

jianhong committed:
add filter condition for prepare_counts.

jianhong committed:
fix a typo in homer_compartments

nf-core-bot committed:
Template update for nf-core/tools version 2.5

nf-core-bot committed:
Template update for nf-core/tools version 2.5.1

nf-core-bot committed:
Template update for nf-core/tools version 2.6

jianhong committed:
change the output folder for test_hipeak.conf

jianhong committed:
fix a bug if juicer_tools_jar is false, there is no JUICER_PRE.out

jianhong committed:
Merge remote-tracking branch 'upstream/nf-core-template-merge-2.6' into dev2

jianhong committed:
fix lint

jianhong committed:
prettier write

jianhong committed:
fix the space issue for enzyme_cut.nf

jianhong committed:
fix multiple lint issue

jianhong committed:
reformat modules.config;
add merge interactions module;

jianhong committed:
update compartment, tad, apa tools.

jianhong committed:
add hicdcplus interaction caller.

jianhong committed:
update more differential methods.
update more v4c.
update more loop calling.

jianhong committed:
update v4c and circos plot.

jianhong committed:
fix an issue of hicdcplus callloops

jianhong committed:
output channels for debug.

jianhong committed:
add TFEA

jianhong committed:
fix multiple bugs in hicdcplus.

jianhong committed:
update homer install.nf

jianhong committed:
update homer install.nf

nf-core-bot committed:
Template update for nf-core/tools version 2.7.1

jianhong committed:
Merge remote-tracking branch 'upstream/nf-core-template-merge-2.7.1' into dev2

jianhong committed:
update documentation and modules.

jianhong committed:
remove local_modules test.

jianhong committed:
update documentations

jianhong committed:
fix the white space issue.

jianhong committed:
fix a typo in eigenvector.nf

jianhong committed:
add peakachu.

jianhong committed:
run python black

jianhong committed:
fix the read_csv error for MAPS.py

jianhong committed:
update annotations from 5k,10K bins to R2 peaks; fix peakachu.

jianhong committed:
remove the usage.html

jianhong committed:
chmod peakachu

jianhong committed:
fix peakachu.

nf-core-bot committed:
Template update for nf-core/tools version 2.7.2

jianhong committed:
adjust the output folder

jianhong committed:
add tools input check.

jianhong committed:
Merge remote-tracking branch 'upstream/nf-core-template-merge-2.7.2' into dev2

jianhong committed:
fix prettier check

jianhong committed:
fix the duplicated imported module issue.

jianhong committed:
update documentation

jianhong committed:
fix the bugs in atacseqtfea.nf

jianhong committed:
fix the chromosome style for homer tfea analysis;
fix the bug in homer compartments caller.

jianhong committed:
update documentations.

jianhong committed:
update documentation.

jianhong committed:
prettier

jianhong committed:
try to fix a bug in peakachu model.nf

jianhong committed:
fix the difference in seqlevels style when annotation by chippeakanno.

jianhong committed:
handle the errors for chippeakanno.nf

maxulysse commented:
Any possibilities to merge https://github.com/nf-core/hicar/pull/79 ?
The sync to TEMPLATE with tools 2.7.2 would be amazing

maxulysse commented on CHANGELOG.md:
```diff
@@ -3,7 +3,10 @@
-## v1.1.0dev - [05/03/2022]
+## v1.1.0dev - [12/12/2022]
+
+- add support for MseI.
+- add TAD, AB compartments, APA analysis
```
Seems a bit thin for such a big PR

jianhong replied:
Yes. I did development for about half year. :)

maxulysse replied:
Figured as much given the size of the PR.
Would have been easier to review with more tiny PRs.
We usually do big PRs like that only for releases.

jianhong replied:
Yes, it is my fault. But I don't know how to roll it back. Will it be a big concern?

maxulysse replied:
No, don't worry, I'll finish the review.
Try just not to do that again

jfy133 replied:
Maybe mention template update number? Given it seems to be a large part of the update

jfy133 replied:
Looking through it looks like you've added extra tools and stuff, you should list these too - I agree with Maxime this changelog is quite thin...

maxulysse commented:
> Any possibilities to merge #79 ? The sync to TEMPLATE with tools 2.7.2 would be amazing

Never mind, just realized it was included already...

maxulysse commented on nextflow.config:
```diff
@@ -60,40 +89,53 @@ params {
     maps_filter                = 'None'
     maps_model                 = 'pospoisson'
+    maps_3d_ext                = 'sig3Dinteractions.pe.txt'
     merge_map_py_source        = 'https://raw.githubusercontent.com/ijuric/MAPS/91c9c360092b25a217d91b9ea07eba5dd2ac72f4/bin/utils/genomic_features_generator/scripts/merge_map.py'
     feature_frag2bin_source    = 'https://raw.githubusercontent.com/ijuric/MAPS/91c9c360092b25a217d91b9ea07eba5dd2ac72f4/bin/utils/genomic_features_generator/scripts/feature_frag2bin.py'
     make_maps_runfile_source   = 'https://raw.githubusercontent.com/ijuric/MAPS/18c1a337f222130d7c5735d051614e2a253d5319/bin/MAPS/make_maps_runfile.py'
```
why not having the python files in `bin/`?

jianhong replied:
This is an old story. At beginning, I put those files in bin folder. However, during the release v1 review process, they were moved out as a link. By doing this, user can upgrade the scripts by change the url for a quick patch.

maxulysse replied:
I see...
I'll ping the @nf-core/core team to have some feedback

jfy133 replied:
Yeah this is not a good idea

- It's not reproducible - if a user uses a different 'version' of a different script, how does someone trying to reproduce the result know which was used? i.e., they can't just go run hicar themselves based on nf-core instructions  
- Already the repo has been moved: https://github.com/HuMingLab/MAPS, who knows i fthe old one will get archived/deleted (and which point the URLs will not work anymore!)
- Given it's on github  this hsould simply be put into bioconda
- It may not work if someone needs to run on an offline cluster, as the file might be loaded on the compute node (caveat: I've not seen in which module these are used yet)

_at least_ you should copy these to the bin, and modified to include the license in them

jianhong replied:
I will try to copy the scripts soon. The jucier is java bin. Will it be OK to keep as a URL?

maxulysse commented on nextflow.config:
```diff
@@ -60,40 +89,53 @@ params {
+    hicdcplus_cutoff_fdr       = 0.05
+
+    // peakachu
+    peakachu_pretrained_url    = 'http://3dgenome.fsm.northwestern.edu/peakachu/HiCAR-models/HiCAR-peakachu-pretrained.'
```
Is this URL correct?

jianhong replied:
This url were provided by the peakachu authors. We have close collaboration with them and they generated the classifier for us.

jfy133 replied:
They should put peakachu on bioconda (see comments above)

They already provide conda instructionS: https://github.com/tariks/peakachu! It would be trivial to put onto bioconda itself!

jianhong replied:
I am not familiar with this process. Is there any quick documentation for that. I will forward the doc to the authors. Another thing, there is another tool also called peakachu in bioconda. Will it be a concern?

maxulysse commented on nextflow.config:
```diff
@@ -111,7 +154,12 @@ params {
     // Juicer Java paramaters
     juicer_jvm_params          = '-Xms512m -Xmx4096m'
-    juicer_tools_jar           = 'https://github.com/aidenlab/Juicebox/releases/download/v2.13.06/juicer_tools_2.13.06.jar'
+    juicer_tools_jar           = 'https://github.com/aidenlab/Juicebox/releases/download/v2.16.00/juicer_tools_2.16.00.jar'
+    juicer_box_jar             = 'https://github.com/aidenlab/Juicebox/releases/download/v2.16.00/juicebox_2.16.00.jar'
```
Can't that be in `bin/` or somewhere else?

jianhong replied:
Save as above. I think this can be upgraded to nearest release version. I will change the url to new version by next push.

jfy133 replied:
See my comments above, this is not stable: the tools are being split up: https://github.com/aidenlab/Juicebox#breaking-news

so it may not be clear in the future which release should be used etc.

maxulysse commented on subworkflows/local/compartments.nf:
```diff
@@ -0,0 +1,85 @@
+    ch_annotation_files     = Channel.empty()
+    ch_compartments         = Channel.empty() // a bigwig files channel
+
+    switch(params.compartments_tool){
+        case "cooltools":
```
I llike this switch case

jfy133 replied:
Agreed!

maxulysse commented on subworkflows/local/interaction_caller/hicexplorer.nf:
```diff
@@ -0,0 +1,23 @@
+    matrix                          // [ bin_size, [cool] ] corrected matrix file
+
+    main:
+    ch_multiqc_files = Channel.empty()
```
Not sure I follow the logic with this `ch_multiqc_files`>
Can you explain the idea behind that?

jianhong replied:
I leave it as standard output for all downstream analysis. Sorry I don't have time to implement them yet but they in TODO list.

jfy133 replied:
If you don't mix anything into it, then I would remove them until you need to d it - I keep noticing this and wanting to leave comments that you're forgetting to mix in multiqc output files...

jfy133 replied:
Or at least leave comments saying '// Adding in later release' or similar

jianhong replied:
This will be fixed during the reviewer process.

maxulysse commented on subworkflows/local/tfea/homer.nf:
```diff
@@ -0,0 +1,36 @@
+    main:
+    ch_versions = Channel.empty()
+
+    if(!workflow.containerEngine){
```
So I'm guessing, not working with conda

jianhong replied:
I need to double check about this. The homer results will be different if there is no pre-installed genome files (the pipeline will try to install the files). For containers, the pipeline will use custom genomes and annotation files 'on-the-fly' because the homer folder are not writable and which in turn, the sequence name will keep consistent.

jfy133 commented on .gitignore:
```diff
@@ -6,5 +6,4 @@ results/
 testing*
 *.pyc
-output/
-null/
```
I think it's safer to include these two still, but up to you

jianhong replied:
I did not noticed that they are removed. I will put it back.

jfy133 commented on assets/multiqc_config.yml:
```diff
@@ -3,9 +3,153 @@ report_comment: >
+    parent_name: "PAIRS QC"
+    parent_description: "Pairs QC report"
+    section_name: "PAIRS reads stats"
+    description: 'Table of mapped reads stats. The duplication rate should be lower than 50%. 20%~30% is a good level for 60M PE reads for human genome. For shallow seq, <30% duplication rate should be the cutting value for further deep sequence. Too low duplication rate (<5%) may indicate low specific chromatin interacts. Those number is estimated by several hypothesis. We set the actual mappabl ...
```
```suggestion
    description: 'Table of mapped reads stats. The duplication rate should be lower than 50%. 20%~30% is a good level for 60M PE reads for the human genome. For shallow sequencing, <30% duplication rate should be the cutting value for further deep sequencing. Too low duplication rate (<5%) may indicate low specific chromatin interactions. These numbers are estimated by several hypothesis. We set the actual mappable genome size to 70% of the original genome size. And the open chromatin interacts region is about 2% of the genome. Take the human genome as an example, the possible unique reads for 60M will be, (3G * 70% * 2%)/60M, about 70%. This is about a 30% duplication rate. Those hypotheses are consistent with real data of ChIP-seq<a href="https://pubmed.ncbi.nlm.nih.gov/24782889/"></a>.'
```

jfy133 commented on assets/multiqc_config.yml:
```diff
@@ -3,9 +3,153 @@ report_comment: >
+  other_side_genomic_dist:
+    parent_id: "pairs_qc"
+    section_name: "The genomeic element distribution for the peaks interacting with promoters"
+    description: "The genomic element distribution of the other sites of promoter peaks. Promoter region is defined as upstream 2000 and downstream 500 from TSS. geneDownstream is defeined as downstream 2000 from TES."
```
```suggestion
    description: "The genomic element distribution of the other sites of promoter peaks. Promoter region is defined as upstream 2000 and downstream 500 from TSS. geneDownstream is defined as downstream 2000 from TES."
```

jfy133 commented on assets/multiqc_config.yml:
```diff
@@ -3,9 +3,153 @@ report_comment: >
+  pairs_peaks_genomic_dist:
+    parent_id: "pairs_qc"
+    section_name: "Peaks genomic element distribution"
+    description: "The genomic element distribution of the peaks. Promoter region is defined as upstream 2000 and downstream 500 from TSS. geneDownstream is defeined as downstream 2000 from TES."
```
```suggestion
    description: "The genomic element distribution of the peaks. Promoter region is defined as upstream 2000 and downstream 500 from TSS. geneDownstream is defined as downstream 2000 from TES."
```

jfy133 commented on bin/MAPS.py:
```diff
@@ -16,6 +16,7 @@
 # ## 7. Fix the wrong data type when reading chromosome name as number
 # ## 8. add weight for counts in case the input file is a count ready file
 # ## 9. fix the indent space
+# ##10. add count for inter-chromosomes, and fix the filenames
```
The header is missing a license! At the very minimum, Please include this in the header - this is very important!

jianhong replied:
one question about the license, if the original license is GPL-3.0, should I keep GPL-3.0 or change to MIT?

jfy133 replied:
You must keep the original! Read the license if you're not sure.

jianhong replied:
Got it.

jfy133 commented on docs/usage.md:
```diff
@@ -61,21 +61,81 @@ TREATMENT,3,AEG588A6_S6_L004_R1_001.fastq.gz,AEG588A6_S6_L004_R2_001.fastq.gz,,
+
+### Call high resolution interactions
+
+The high resolution interaction caller are also available for confident transcription factor enrichment analysis. However, please note that the high resolution interaction caller are not bin-based but peak based analysis, which ask high computational resources.
```
```suggestion
The high resolution interaction caller is also available for confident transcription factor enrichment analysis. However, please note that the high resolution interaction caller are not bin-based but peak based analysis, which uses high computational resources.
```

jfy133 commented on docs/usage.md:
```diff
@@ -61,21 +61,81 @@ TREATMENT,3,AEG588A6_S6_L004_R1_001.fastq.gz,AEG588A6_S6_L004_R2_001.fastq.gz,,
+
+In downstream, differential analysis available for called interactions. Available tools are Bioconductor packages such as `edgeR`, and `diffhic`, and [`HiCExplorer`](https://hicexplorer.readthedocs.io/en/latest/). We borrowed capture Hi-C analysis pipeline from HiCExplorer to do the differential analysis. Different from `edgeR` and `diffhic` pipeline, HiCExplorer pipeline does not require the rep ...
+
+For annotation, we will use Bioconductor package [`ChIPpeakAnno`](https://bioconductor.org/packages/ChIPpeakAnno/). Please note that, the involved genes are not only distance based annotation. The most of the interaction calling tools are bin-based caller, and the bin size are kilo-base or even more, which make the annotation difficult. For HiCAR data, the R2 reads are Tn5 insertion site of the o ...
```
```suggestion
For annotation, we will use Bioconductor package [`ChIPpeakAnno`](https://bioconductor.org/packages/ChIPpeakAnno/). Please note that, the involved genes are not only distance based annotation. The most of the interaction calling tools are bin-based caller, and the bin size are kilo-base or even more, which make the annotation difficult. For HiCAR data, the R2 reads are Tn5 insertion site of the open chromatin. And most of the R2 reads will be an anchor of annotion for the gene promoters. We will annotate the interactions by the annotation of called ATAC (R2) peaks located within the interaction regions.
```

jfy133 commented on docs/usage.md:
```diff
@@ -172,7 +232,19 @@ Command output:
-To bypass this error you would need to find exactly which resources are set by the `MAPS_CUT` process. The quickest way is to search for `process MAPS_CUT` in the [nf-core/hicar Github repo](https://github.com/nf-core/hicar/search?q=process+MAPS_CUT). We have standardised the structure of Nextflow DSL2 pipelines such that all module files will be present in the `modules/` directory and so based o ...
+#### For beginners
+
+A first step to bypass this error, you could try to increase the amount of CPUs, memory, and time for the whole pipeline. Therefor you can try to increase the resource for the parameters `--max_cpus`, `--max_memory`, and `--max_time`. Based on the error above, you have to increase the amount of memory. Therefore you can go to the [parameter documentation of hiar](https://nf-co.re/hicar/1.0.0/para ...
```
```suggestion
A first step to bypass this error, you could try to increase the amount of CPUs, memory, and time for the whole pipeline is allowed to access. Therefore you can try to increase the resource limits for the parameters `--max_cpus`, `--max_memory`, and `--max_time`. Based on the error above, you have to increase the amount of memory. Therefore you can go to the [parameter documentation of hiar](https://nf-co.re/hicar/1.0.0/parameters) and scroll down to the `show hidden parameter` button to get the default value for `--max_memory`. In this case 128GB, you than can try to run your pipeline again with `--max_memory 200GB -resume` to skip all processes that were already completed. If you cannot increase the resource of the complete pipeline, you can try to adapt the resource for a single process as mentioned below.
```

jfy133 commented on docs/usage.md:
```diff
@@ -172,7 +232,19 @@ Command output:
+
+#### Advanced option on process level
+
+To bypass this error you would need to find exactly which resources are set by the `MAPS_CUT` process. The quickest way is to search for `process MAPS_CUT` in the [nf-core/hicar Github repo](https://github.com/nf-core/hicar/search?q=process+MAPS_CUT).
```
You can also refer to: https://nf-co.re/docs/usage/configuration#tuning-workflow-resources insead

jianhong replied:
add this to beginner section.

jfy133 commented on docs/usage.md:
````diff
@@ -290,3 +370,21 @@ The error may caused by out of memory (although the error message seems to be un
+process.executor = 'slurm' // the workload manager name
+process.queueSize = 10 // the job queue size
+process.clusterOptions = "-J nextFlowHiCAR -p scavenger" // the options, here -p request a specific partition for the resource allocation. It will be different in your cluster.
+```
````
You should again leave this to nf-core level documentation: https://nf-co.re/docs/usage/configuration#tuning-workflow-resources

jfy133 commented on modules/local/bioc/atacseqtfea.nf:
```diff
@@ -0,0 +1,180 @@
+    #######################################################################
+    #######################################################################
+    ## Created on Aug. 24, 2021 for ATACseqTFEA
+    ## Copyright (c) 2021 Jianhong Ou (jianhong.ou@gmail.com)
```
I would again include the MIT license in all of these custom scripts.

Also why are these embedded in processes and not in `bin/` like others? I think it would make sense to be  consistent.

jianhong replied:
This is a good question. Indeed I agree with that all the scripts should be put in bin/ folder. However, all the script I wrote were moved out to follow the new style of customized module ( I don't know where I got this idea, but I take lots of time to make this change). I will put the MIT license to the header.

jfy133 commented on modules/local/circos/circos_prepare.nf:
```diff
@@ -35,11 +35,10 @@ process CIRCOS_PREPARE {
 
     ## Options
-    ## make_option(c("-i", "--interaction"), type="character", default=NULL, help="interaction bedpe file", metavar="string")
+    ## make_option(c("-i", "--interactions"), type="character", default=NULL, help="interaction bedpe file", metavar="string")
```
This looks like it was a bug that you fixed (correct me if I'm wrong) - it would be important report fixes bugs in the CHANGELOG too

jianhong replied:
will do.

jfy133 commented on modules/local/cooler/balance/main.nf:
```diff
@@ -0,0 +1,34 @@
+process COOLER_BALANCE {
```
You should not have a custom local module when nf-core/modules exist!

https://nf-co.re/modules/cooler_balance

If you need to tweak it you can include patches with `nf-core modules diff`

jfy133 commented on modules/local/cooltools/eigscis.nf:
```diff
@@ -0,0 +1,37 @@
+process COOLTOOLS_EIGSCIS {
```
This should be an official nf-core module

jfy133 commented on modules/local/cooltools/plotnpz.nf:
```diff
@@ -0,0 +1,64 @@
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${meta.id}_${meta.bin}"
+    """
+    #!/usr/bin/env python
```
This is a bit confusing - it seems to be a custom python script but still uses a cooltools container, is that correct?

jianhong replied:
Yes. It will be used to plot the data. The plotting tools are based on cooltools.

jfy133 commented on modules/local/cooltools/virtual4c.nf:
```diff
@@ -0,0 +1,56 @@
+    def args = task.ext.args ?: ''
+    def prefix = task.ext.prefix ?: "${bin_size}"
+    """
+    # VIEWPOINT : the viewpoint to use for the virtual 4C profile. Provide as a UCSC-string (e.g. chr1:1-1000)
```
Same again here. Don't use an 'officlal' nf-core/modules naming scheme if you've written custom scripts as this could be confusing for other developers/users

jianhong replied:
I will change the names.

jfy133 commented on modules/local/hicexplorer/chicaggregatestatistic.nf:
```diff
@@ -0,0 +1,33 @@
+process HICEXPLORER_CHICAGGREGATESTATISTIC {
```
Should be a proper nf-core module

jfy133 commented on modules/local/peakachu/score_genome.nf:
```diff
@@ -0,0 +1,36 @@
+process PEAKACHU_SCORE {
+    label 'process_single'
+
+    conda "bioconda::cooltools=0.5.2"
```
Is this the correct container?

jianhong replied:
In current stage, yes. If we got peakachu container, I will change it.

jfy133 commented on nextflow.config:
```diff
@@ -60,40 +89,53 @@ params {
+    peakachu_pretrained_url    = 'http://3dgenome.fsm.northwestern.edu/peakachu/HiCAR-models/HiCAR-peakachu-pretrained.'
+
+    // HiCExplorer
+    peakinteractionsthreshold  = 2
```
Given your other parmeters, shouldn't this be snake case format?

```suggestion
    peak_interactions_threshold  = 2
```

Would need to be updated elsehwere

jianhong replied:
fixed.

jfy133 commented on nextflow.config:
```diff
@@ -211,29 +278,30 @@ process.shell = ['/bin/bash', '-euo', 'pipefail']
-    version = '1.1.0dev'
+    nextflowVersion = '!>=22.10.1'
+    version         = '1.1.0dev'
+    doi             = ''
```
You should add the DOI to your paper here, this way it'll be inserted into the new MultiQC methods description :)

jianhong replied:
added.

jfy133 commented on nextflow_schema.json:
```diff
@@ -150,6 +150,16 @@
+                },
+                "publish_mappability": {
+                    "type": "boolean",
+                    "description": "publish mappability",
```
Missing descirption and helptex here, what files are you actually publishing? Where are they going, in `results/` or publishing somewhere else?

jfy133 commented on nextflow_schema.json:
```diff
@@ -338,6 +348,140 @@
+                    "description": "FDR cutoff value",
+                    "fa_icon": "fas fa-percent",
+                    "hidden": false,
+                    "help_text": "Minimal cutoff FDR value in the called connections. See help from [HiCDCPlus documentation](https://bioconductor.org/packages/HiCDCPlus/), default 0.05.",
```
nice to point to the tool docs :+1:

jfy133 commented on nextflow_schema.json:
```diff
@@ -338,6 +348,140 @@
+                },
+                "tad_tool": {
+                    "type": "string",
+                    "description": "The caller for TADs, available choices: 'cooltools', 'hicexplorer', 'homer'",
```
For ALL parameters that have choices, do not list these i nthe description, use the pipe-separated `enum` option in the schema build

(`nf-core schema build` -> scroll to parameter -> press cog on the far right of parameter)

jfy133 commented on nextflow_schema.json:
```diff
@@ -338,6 +348,140 @@
+                },
+                "juicer_box_jar": {
+                    "type": "string",
+                    "default": "https://github.com/aidenlab/Juicebox/releases/download/v2.16.00/juicebox_2.16.00.jar",
```
If you stick with this, you _definitely_ need a description/help_text of what this file should be, where to get from etc.

jianhong replied:
I will change this later.

jfy133 commented on subworkflows/local/apa.nf:
```diff
@@ -0,0 +1,64 @@
+include { HICEXPLORER_HICAGGREGATECONTACTS } from '../../modules/local/hicexplorer/hicaggregatecontacts'
+include { JUICER_APACALLER                 } from './apa/juicer'
+include { COOLTOOLS_APACALLER              } from './apa/cooltools'
+workflow APA {
```
```suggestion

workflow APA {
```

jfy133 commented on subworkflows/local/apa.nf:
```diff
@@ -0,0 +1,64 @@
+    take:
+    matrix            // tuple val(meta), path(cool/hic), signal file
+    peaks             // path(1d/2d peaks) for APA analysis
+    additional_param
```
What is this? If it's additional parameters that a user can supply, this should come via `ext.args` not an input parameter

jfy133 commented on subworkflows/local/apa/cooltools.nf:
```diff
@@ -0,0 +1,21 @@
+    peaks                  // path(1D peaks)
+
+    main:
+    COOLTOOLS_PILEUP(matrix, peaks).npz | COOLTOOLS_PLOTNPZ
```
We generally don't use the nextflow pipe in nf-core (I think @maxulysse can oyu confim?, I would opt to split over two lines and call the explicit output of PILEUP into PLOTNPZ (again, speeds up reviews

But this is just a  stylistic thing

jfy133 commented on subworkflows/local/apa/juicer.nf:
```diff
@@ -0,0 +1,37 @@
+    additional_param       // values [juicer_box_jar, chromsize, bin, merged_loops]
+
+    main:
+    //additional_param.view()
```
```suggestion
```

jianhong replied:
will remove all the viewers.

jfy133 commented on subworkflows/local/apa/juicer.nf:
```diff
@@ -0,0 +1,37 @@
+
+    //ch_hic_loops.view{"loops: $it"}
+    //peaks.view{"peak: $it"}
+    //additional_param.map{[it[0], it[1]]}.view()
```
```suggestion
```

jfy133 commented on subworkflows/local/callatacpeak.nf:
```diff
@@ -2,28 +2,33 @@
+    UCSC_BEDGRAPHTOBIGWIG
+        as UCSC_BEDGRAPHTOBIGWIG_PER_GROUP;
+    UCSC_BEDGRAPHTOBIGWIG
+        as UCSC_BEDGRAPHTOBIGWIG_PER_SAMPLE } from '../../modules/nf-core/ucsc/bedgraphtobigwig/main'
```
I find this much less readable personally... I would stick to one line per `include` (but again stylisitc thing)

jfy133 commented on subworkflows/local/circos.nf:
```diff
@@ -7,15 +7,15 @@ include { CIRCOS                    } from '../../modules/local/circos/circos'
     //create circos config
-    //input=path(bedpe), val(ucscname), path(gtf), path(chromsize)
+    //input=val(meta), path(bedpe), val(ucscname), path(gtf), path(chromsize)
     ch_version = CIRCOS_PREPARE(bedpe.combine(ucscname).combine(gtf).combine(chromsize)).versions.first()
```
Just want to check `.combine` is the right function you want to use here, if you have multiple elements per channel you're combining, this will make an all-v-all set of 'pairs'; are you sure this isn't meant to be e.g. `join` so it's only executed on a per-sample basis?

jfy133 commented on subworkflows/local/compartments.nf:
```diff
@@ -0,0 +1,85 @@
+    igv             = ch_track_files
+    anno            = ch_annotation_files
+    versions        = ch_versions          // channel: [ versions.yml ]
+    mqc             = ch_multiqc_files
```
I dn't see any mixing into here, you can remove this and the channel createion at the top if you wish

jfy133 commented on subworkflows/local/compartments_caller/hicexplorer.nf:
```diff
@@ -0,0 +1,49 @@
+    HICEXPLORER_HICTRANSFORM(
+        cool
+    )
+    ch_version = ch_version.mix(HICEXPLORER_HICTRANSFORM.out.versions.ifEmpty(null))
```
Are you sure you want the ifEmpty everywhere? You are definitely executing the module - if it doesn't produce a version this indicates a tool error for some reason... here you could be hiding such an error by saying if empty `[]`

jianhong replied:
`ifEmpty(null)` will be removed.

jfy133 commented on subworkflows/local/cooler.nf:
```diff
@@ -35,24 +37,33 @@ workflow COOLER {
-        ch_version = ch_version.mix(JUICER.out.versions)
+        JUICER_PRE(DUMPREADS_PER_GROUP.out.gi, juicer_tools_jar, chromsizes, juicer_jvm_params)
+        ch_hic = JUICER_PRE.out.hic
+        ch_version = ch_version.mix(JUICER_PRE.out.versions)
```
Given you provide a default, it will always be present - why the `if` statement?

jianhong replied:
It will be removed.

jfy133 commented on subworkflows/local/hipeak.nf:
```diff
@@ -1,80 +1,78 @@
+        short_bed_postfix
+    )
+    ch_versions = ch_versions.mix(R1_PEAK.out.versions.ifEmpty(null))
+    //ch_trackfiles = ch_trackfiles.mix(
+    //    R1_PEAK.out.bws.map{[it[0].id+"_R1",
+    //        RelativePublishFolder.getPublishedFolder(workflow,
+    //            'UCSC_BEDGRAPHTOBIGWIG_PER_R1_GROUP')+it[1].name]})
```
Please delete all commented out code in this subwf

jfy133 commented on subworkflows/local/interaction_caller/homer.nf:
```diff
@@ -0,0 +1,29 @@
+/*
+ * Call interaction peaks by Homer, current not proper
```
What does `current not proper` refer to?

jianhong replied:
This caller will be removed.

jfy133 commented on subworkflows/local/interaction_caller/maps.nf:
```diff
@@ -0,0 +1,54 @@
+                                    //    4 scripts: merge_map_py_source,
+                                    //    5 scripts: feature_frag2bin_source
+                                    //    6 scripts: make_maps_runfile_source
+                                    // ]
```
This is quite bad practise to pass as an additional param, if these are important you use the nf-core `meta` for the values, and update modules to have explicit channels if required and thus passed directly to the subworkflow

jfy133 commented on subworkflows/local/interaction_caller/maps_multienzyme.nf:
```diff
@@ -0,0 +1,65 @@
+/*
+ * Createing Genomic Features Files for MAPS
```
```suggestion
 * Createing Genomic Features Files for MAPS
```
```suggestion
 * Creating Genomic Features Files for MAPS
```

jfy133 commented on subworkflows/local/interaction_caller/maps_multienzyme.nf:
```diff
@@ -0,0 +1,65 @@
+        ch_version = Channel.empty()
+        ch_digest = cool_bin.combine(Channel.fromPath(params.maps_digest_file))
+        MAPS_FEND(ch_digest, chrom_sizes)
+    }else{
```
```suggestion
    } else {
```

And for the reast of this file

jfy133 commented on subworkflows/local/interactions.nf:
```diff
@@ -0,0 +1,93 @@
+// Call interaction loops
+// input:
+// [[2D signals], [1D peaks], [additional_files]]
+// Ask all the output of each module must contain channel:
```
```suggestion
// The output of each module must contain channel:
```

jfy133 commented on subworkflows/local/tads.nf:
```diff
@@ -0,0 +1,76 @@
+//
+// Call TADs
+// Ask all the output of each module must contain channel:
```
```suggestion
// All the output of each module must contain channel:
```

jfy133 commented on subworkflows/local/tads.nf:
```diff
@@ -0,0 +1,76 @@
+            ch_versions = COOLTOOLS_TADS.out.versions
+            ch_circos_files = COOLTOOLS_TADS.out.tads.map{[it[0], it[2]]}
+            break
+        case "hicexploer":
```
```suggestion
        case "hicexplorer":
```

Maybe check for similar typos in the repo

jfy133 commented on subworkflows/local/tads_caller/homer.nf:
```diff
@@ -0,0 +1,30 @@
+    genome       // value: UCSC genome name
+
+    main:
+    ch_multiqc_files = Channel.empty()
```
Remove if not used

jfy133 commented on workflows/hicar.nf:
```diff
@@ -38,29 +39,73 @@ ch_anchor_peaks = Channel.empty()
+}
+if (!params.tfea_tool in ['atacseqtfea', 'homer']){
+    exit 1, 'The tfea_tool must be one of atacseqtfea, or homer'
+}
```
This can all be removed once you update the JSON schema with the enums!

jianhong replied:
fixed.

jfy133 commented on workflows/hicar.nf:
```diff
@@ -232,226 +313,403 @@ workflow HICAR {
-            PREPARE_GENOME.out.gtf,
-            params.r1_pval_thresh
+    if(checkToolsUsedInDownstream('hicexplorer', params)){
+        // HICEXPLORER_HICCORRECTMATRIX(COOLER.out.raw) // not work, buggy
```
Remove if it doesn't work

jianhong replied:
removed.

jfy133 commented on workflows/hicar.nf:
```diff
@@ -232,226 +313,403 @@ workflow HICAR {
+    //
+    if(checkToolsUsedInDownstream('cooltools', params)){
+        if(params.compartments_tool == 'cooltools'){
+            ch_comp_matrix = COOLER.out.mcool // colltools ask the resolution match the tiled genome
```
```suggestion
            ch_comp_matrix = COOLER.out.mcool // cooltools ask the resolution match the tiled genome
```

jfy133 commented on workflows/hicar.nf:
```diff
@@ -232,226 +313,403 @@ workflow HICAR {
+                )
+                ch_versions = ch_versions.mix(V4C.out.versions.ifEmpty(null))
+            }else{
+                // NOT applicable
```
I would remove the `else` then, just keep with if, or come up with a reasonable 'else' action (maybe just `null` or print nothing)

jianhong replied:
removed.

jfy133 commented on workflows/hicar.nf:
```diff
@@ -232,226 +313,403 @@ workflow HICAR {
+            }
+        }else{
+            //ch_v4c_files.view()
+            //COOLER.out.cool.view()
```
```suggestion
```

might be worht doing a search through the entire pipeline for `.view` and remove all related lines

jianhong replied:
removed.

jfy133 reviewed (changes requested):
So overall good job! I particularly like the new usage descriptions - they very detailed and nice! :+1: 

This is clearly a very complex pipeline, however I share some large concerns with Maxime related to the large number of custom scripts and local modules.

At the bare minimum, for this release:

- You must include the license in **all** custom scripts in `bin/` and these should ideally match that of the pipeline: MIT to prevent copyrighting  issues, also particularly given you have copied and modified existing scripts in some cases.
- You should ideally copy all the of the URL versions of scripts also into `bin/` and not allow a user to update these (as this conficts with the reproducibility and portability components of the nf-core requirements). I would suggest checking with the authors of those scripts this is OK (and make sure these also have the license in the headers etc.)

However for the next release I would expect that:

- You have a many local versions of modules where an offiical nf-core/module exist, these should be switched out (particularly cooler) with the official version (and modified with `diff` if necessary)
- You have many local versions of tools these should converted to official nf-core/modules (cooltool, explorer etc.)

In addition to making it cleaner,  this will massively reduce the revieiwing load at a pipeline level so you'll get faster release reviews in the future. A reviewer can assume an official nf-core module has been reviewed already at the module level and can skip reviewing it at the pipeline level. In this regard,  I've opted to skip reviewing the local modules in this review this... I will just trust that your local modules 'work' correctly.. 

For some of the local modules using local scripts I owuld highly recommend investing time and effort into helping the authors of those tools into putting them properly on bioconda - this will also make the pipeline more portable and reproducible.

Finally: I left a comment somewhere about leaving spaces before/after `else`, I realise your style is without and you've done that everywhere and that's fine - please ignore that comment!

jianhong committed:
Update assets/multiqc_config.yml

Co-authored-by: James A. Fellows Yates <jfy133@gmail.com>

jianhong committed:
fix a bug in recenterpeak.nf for bed format file.

jianhong committed:
Merge branch 'dev2' of github.com:jianhong/hicar into dev2

jianhong committed:
Update docs/usage.md

Co-authored-by: James A. Fellows Yates <jfy133@gmail.com>

jianhong committed:
Update docs/usage.md

Co-authored-by: James A. Fellows Yates <jfy133@gmail.com>

jianhong committed:
Update docs/usage.md

Co-authored-by: James A. Fellows Yates <jfy133@gmail.com>

jianhong committed:
update documentation.

jianhong committed:
add license to customized scripts.

jianhong committed:
rename the tools.

jianhong committed:
move the tools check step to nextflow schema.
remove the java memory parameter.
remove the unused scripts.

jianhong committed:
remove the ifEmpty for versions.
replace combine by join

jianhong committed:
remove if statement for juicer_tools_jar check.

jianhong committed:
remove homer interaction caller.

jianhong committed:
Update subworkflows/local/interaction_caller/maps_multienzyme.nf

Co-authored-by: James A. Fellows Yates <jfy133@gmail.com>

jianhong committed:
Update workflows/hicar.nf

Co-authored-by: James A. Fellows Yates <jfy133@gmail.com>

jianhong committed:
removed un-used code.

jianhong committed:
add doi to nextflow.config file.

jianhong committed:
mark todo for mqc channel.

jianhong committed:
change local cooler modules to nf-core cooler modules.

jianhong committed:
Update subworkflows/local/interactions.nf

Co-authored-by: James A. Fellows Yates <jfy133@gmail.com>

jianhong committed:
Update subworkflows/local/tads.nf

Co-authored-by: James A. Fellows Yates <jfy133@gmail.com>

jianhong committed:
fix typo of hicexplorer

jianhong committed:
prittier fix the format of nextflow_schema.json

jianhong committed:
reformat peakachu_score_genome.py by black.

jianhong committed:
Update subworkflows/local/tads.nf

Co-authored-by: James A. Fellows Yates <jfy133@gmail.com>

jianhong committed:
add chicQualityControl to filter reference points.

jianhong committed:
fix the channel re-use issue for juicebox.

jianhong committed:
fix hpc usage documentation for slurm job queue size.

jianhong committed:
download the scripts of MAPS to bin folder;
fix the wrong chr2 columns in MAPS calling;
update the juicer tools to release V3.

jianhong committed:
black bin fix for maps scripts.

jianhong committed:
skip error for juicer compartment caller.

jianhong committed:
change the output folder of MAPS from MAPSpeak to maps;
export juicer normalization method as a parameter;
generate bigwig file for juicer compartment results.

jianhong committed:
fix the output folder for juicer bigwig file.
export bigwig file for homer compartment caller.

jianhong committed:
update ChangeLogs.

jianhong committed:
update ChangeLog.md

jianhong commented:
> * You must include the license in **all** custom scripts in `bin/` and these should ideally match that of the pipeline: MIT to prevent copyrighting  issues, also particularly given you have copied and modified existing scripts in some cases.

> * You should ideally copy all the of the URL versions of scripts also into `bin/` and not allow a user to update these (as this conficts with the reproducibility and portability components of the nf-core requirements). I would suggest checking with the authors of those scripts this is OK (and make sure these also have the license in the headers etc.)

I copied MAPS scripts into `bin/` with the original license. And add or license to the header of scripts in `bin/`.


> * You have a many local versions of modules where an offiical nf-core/module exist, these should be switched out (particularly cooler) with the official version (and modified with `diff` if necessary)

I updated multiple cooler modules to official nf-core/modules. 

> * You have many local versions of tools these should converted to official nf-core/modules (cooltool, explorer etc.)

I am working on this part, however, it will take time.

jianhong committed:
update pythonBlack

jianhong committed:
fix a typo introduced by rename the parameter for CHICSIGNIFICANTINTERACTIONS

jianhong committed:
fix the out of range issue for compartments caller when generate bigwig files.

jianhong committed:
fix the bedgraph file format output by find_hic_compartments.nf.

jianhong committed:
try to fix the issue of End coordinate bigger than chromosome size.

jianhong committed:
fix the error introduced by assign mismatched seqnames for trimming seqlengths.

jianhong committed:
sort the bedgraph before export by trimbedgraph.nf

jianhong committed:
add scale factor to atac reads coverage cal.

jianhong committed:
update changelog.

jianhong committed:
adjust callpeak resource;
set GET_SCALE output to false;

jianhong committed:
fix the coordinates bug in diffhicar by edger

jianhong committed:
try to fix the sort issue for trimbedgraph.

jianhong committed:
fix the ambiguous redirect error for different version of wc

jianhong committed:
fix the bug of ATAC coverage introduced by scale coverage.

jianhong committed:
fix the typo

jianhong committed:
try to fix the sortbed issue for trimbegraph.nf

jianhong committed:
fix the negative width value for trimbedgraph.

jianhong committed:
update the circos plot.

jianhong committed:
fix the indent issue.

jianhong committed:
fix the bug that peakachu_score_genome only output one chromosome.
update circos.conf.

jianhong committed:
update reads_summary table to include unmapped/multiple mapped info.

jianhong committed:
add the possibility to subsample to balance the input reads

jianhong committed:
update circos plot
update hicexplorer tad plot

jianhong committed:
update circos plots for homer

jianhong committed:
keep standard chromosome only for hicdcplus.

jianhong committed:
change the cutadapt error tolerance from 0 to 0.15

jianhong committed:
fix the issue of dot in group name for APA plot.

jianhong committed:
replace the dots by '_' in the samples name

jianhong committed:
Prettier check

jianhong committed:
fix the bug in merge process if no interaction called.

jianhong committed:
fix the sort method for juicer pre.

jianhong committed:
change the default normalization method of juicer tools from KR to SCALE.

jianhong committed:
update cooltools version;
fix a bug in chromsome name filter for hipeak.

jianhong committed:
fix the bug that circos can not handle emsembl style chromosome names.

jianhong committed:
fix the bug that circos can not handle ensembl style chromosome names.

jianhong committed:
add more annotation plots.

jianhong committed:
fix the spaces in last update.

jianhong committed:
adjust parameters for cooltools/insulation

jianhong committed:
loosing hipeak calling condition.

nf-core-bot committed:
Template update for nf-core/tools version 2.8

jianhong committed:
merge  upstream/nf-core-template-merge-2.8

jianhong committed:
fix the issue `TypeError: float() argument must be a string or a real number, not 'collections.OrderedDict'` introduced by multiQC.

jianhong committed:
update modules;
comment out registry to quay.io until the bugs fixed for gunzip and 
untar.

jianhong committed:
adjust the pipeline controlers;
update the docker.registry.
